Add support for `--create-only` migrations

shayneczyzewski opened the pull request:
# Description

Adds support for `prisma db migrate --create-only` migrations. This is necessary when Prisma cannot automatically create a migration for you based on your schema changes.

This required a few additional changes:
- The CLI invocation for `db migrate-dev` was extended to use options for `--name foo` and `--create-only`
- The copy file draft clears out the contents of the destination directory (not strictly related, but discovered here that it should have behaved that way)
- Changed the way we handle the last concurrence checksum file indicating the user is in sync. Now we must also compare the migrations dir, not just the schema.

Fixes #798

## Type of change

Please select the option(s) that is more relevant.

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
 - change CLI for invoking with migration name from `wasp db migrate-dev foo` to `wasp db migrate-dev --name foo`
- [x] This change requires a documentation update

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -60,3 +66,42 @@ dbSchemaChecksumOnLastGenerateFileInDbRootDir = [relfile|schema.prisma.wasp-gene
+    go :: [String] -> MigrateArgs -> MigrateArgs
+    go ("--create-only" : rest) mArgs = go rest mArgs {_isCreateOnlyMigration = True}
+    go ("--name" : name : rest) mArgs = go rest mArgs {_migrationName = Just name}
+    go _ mArgs = mArgs
```
Right now I just ignore and bail for any unexpected arg. Should this `error` instead?

Martinsos replied:
I think error would be better! That way, if they make minor typo, they will know they did a typo, instead of thinking args was applied.

shayneczyzewski replied:
Thanks, agree. I was leaning that way but this was the lazy "should I _realllyyy_ do it?" :D Will do!

Martinsos replied:
hehe sounds good :D!

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Operations.hs:
```diff
@@ -43,26 +49,35 @@ printJobMsgsUntilExitReceived chan = do
   return res
   where
     dbMigrationsDirInProjectRootDir = dbRootDirInProjectRootDir SP.</> dbMigrationsDirInDbRootDir
+    applyLastDbConcurrenceChecksumAction =
+      case lastDbConcurrenceChecksumAction of
+        WriteOnLastDbConcurrenceChecksum ->
+          writeDbSchemaChecksumToFile genProjectRootDirAbs (SP.castFile dbSchemaChecksumOnLastDbConcurrenceFileProjectRootDir)
+        RemoveOnLastDbConcurrenceChecksum ->
+          removeDbSchemaChecksumFile genProjectRootDirAbs (SP.castFile dbSchemaChecksumOnLastDbConcurrenceFileProjectRootDir)
+        IgnoreOnLastDbConcurrenceChecksum -> return ()
 
```
This section got a little more complicated, but hopefully the new type helps. Basically we have some scenarios where we want to write the checksum indicating we are fully in sync with DB (like after a migrate), or remove it when we may have invalidated it (like when we create an empty migration).

When we create an empty migration, we cannot rely on the prisma schema alone to determine if we should migrate, as they can write anything in the new file. That requires our new migrate dir check.

shayneczyzewski commented on waspc/src/Wasp/Generator/FileDraft/CopyDirFileDraft.hs:
```diff
@@ -35,6 +42,8 @@ data CopyDirFileDraft = CopyDirFileDraft
 instance Writeable CopyDirFileDraft where
   write projectRootAbsPath draft = do
     srcDirExists <- doesDirectoryExist $ SP.fromAbsDir srcPathAbsDir
+    dstDirExists <- doesDirectoryExist $ SP.fromAbsDir dstPathAbsDir
+    when dstDirExists $ removeDirectoryRecursive dstPathAbsDir
```
I added this for two reason:
1) It now matches the semantics of copy file, where if it exists it overwrites it and thus they are equal. Now, when we have a dir FD, we know they will be equal (i.e. the destination cannot have "extra" stuff).
2) When a user wants to remove a migration from their wasp project dir, we want that to be removed from the generated project. This may happen when they do a `--create-only` but then don't migrate and just remove the dir. Without this, it would stay in their generated project and the next migrate could apply it.

shayneczyzewski replied:
I don't think this will negatively affect the generation process. The only thing it should do is ensure there is nothing stale leftover, which is what we want. I don't think we have `createDirectory` alone sadly. Prob ok to just use what we have for now I think.

shayneczyzewski commented on waspc/cli/exe/Main.hs:
```diff
@@ -127,8 +127,8 @@ printVersion = do
 dbCli args = case args of
-  ["migrate-dev", migrationName] -> runDbCommand $ Command.Db.Migrate.migrateDev (Just migrationName)
   ["migrate-dev"] -> runDbCommand $ Command.Db.Migrate.migrateDev Nothing
+  "migrate-dev" : migrateArgs -> runDbCommand $ Command.Db.Migrate.migrateDev $ Just migrateArgs
```
In the example above, they would be passed in like: `["--name", "FOO", "--create-only"]`

Martinsos replied:
Ha, this is getting a bit complex now!

Maybe we should just have it accept [String] then, and it can be empty if there are no arguments, instead of Maybe [String]? So [] would play role of Nothing.

Ideally, we would parse the strings and pass some nicely shaped params. That is a bit of work right now though, so I do agree with postponing it. We should probably switch the whole CLI to that applicative library and that would also help with this.

shayneczyzewski replied:
I like your idea of moving the parsing into CLI. I can then pass in the `MigrateArgs` 👍🏻 Will do thnx

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -60,3 +66,42 @@ dbSchemaChecksumOnLastGenerateFileInDbRootDir = [relfile|schema.prisma.wasp-gene
+  deriving (Show, Eq)
+
+emptyMigrateArgs :: MigrateArgs
+emptyMigrateArgs = MigrateArgs {_migrationName = Nothing, _isCreateOnlyMigration = False}
```
rename to defaultMigrateArgs?

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -60,3 +66,42 @@ dbSchemaChecksumOnLastGenerateFileInDbRootDir = [relfile|schema.prisma.wasp-gene
+emptyMigrateArgs :: MigrateArgs
+emptyMigrateArgs = MigrateArgs {_migrationName = Nothing, _isCreateOnlyMigration = False}
+
+asArgs :: MigrateArgs -> [String]
```
AHa, so this transforms them into actual `prisma dev migrate` args?

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -60,3 +66,42 @@ dbSchemaChecksumOnLastGenerateFileInDbRootDir = [relfile|schema.prisma.wasp-gene
+      [["--name", name] | Just name <- [_migrationName migrateArgs]]
+    ]
+
+parseMigrateArgs :: Maybe [String] -> MigrateArgs
```
I would suggest moving this logic into CLI, where the command is parsed.
I would keep `asArgs` here, and I would keep `MigrateArgs` here, but I would move this logic that translates from CLI args to MigrateArgs.
asArgs could maybe go to Jobs/Db though, not 100% sure -> wherever is closer to Prisma.

shayneczyzewski replied:
Sounds good, will do. That should simplify the interface of `Maybe [String]` as well 👍🏻

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -60,3 +66,42 @@ dbSchemaChecksumOnLastGenerateFileInDbRootDir = [relfile|schema.prisma.wasp-gene
+  go migrateArgs emptyMigrateArgs
+  where
+    go :: [String] -> MigrateArgs -> MigrateArgs
+    go ("--create-only" : rest) mArgs = go rest mArgs {_isCreateOnlyMigration = True}
+    go ("--name" : name : rest) mArgs = go rest mArgs {_migrationName = Just name}
```
```suggestion
    go ("--create-only" : rest) mArgs = go rest $ mArgs {_isCreateOnlyMigration = True}
    go ("--name" : name : rest) mArgs = go rest $ mArgs {_migrationName = Just name}
```

I know it is redundant by linter, but for me it greatly increases readibilty, it took me some time to figure out that go takes 2 args here and not 3.

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -60,3 +66,42 @@ dbSchemaChecksumOnLastGenerateFileInDbRootDir = [relfile|schema.prisma.wasp-gene
+  go migrateArgs emptyMigrateArgs
+  where
+    go :: [String] -> MigrateArgs -> MigrateArgs
+    go ("--create-only" : rest) mArgs = go rest mArgs {_isCreateOnlyMigration = True}
```
Niiiceeee, proper fun recursion :D! Yeah this what we are about right :D!?

shayneczyzewski replied:
I know, right! Haskell, baby!!! It has sadly been so long since I wrote a fun little function like this that it took me a second to get my mind into it though lol 😅

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -60,3 +66,42 @@ dbSchemaChecksumOnLastGenerateFileInDbRootDir = [relfile|schema.prisma.wasp-gene
+    go ("--name" : name : rest) mArgs = go rest mArgs {_migrationName = Just name}
+    go _ mArgs = mArgs
+
+data OnLastDbConcurrenceChecksumAction
```
What exactly does this stand for? So on it tells me what was done on the last DB concurrent checksum, or what should be done?

Martinsos replied:
Ah got it know, it should be read as "Write OnLastDbConcurrenceChecksum".
Hmmmmmm.
I don't have much better idea. Maybe a short comment saying that this says what would be done with the checksum file.

Maybe `WriteOnLastDbConcurrentChecksumFile` would be better? So adding File at the end?

Or maybe `Write_OnLastDbConcurrentChecksumFile`?

This `On` is what confuses me, but I guess we can't get rid of it easily as it become name that we use on many places?

shayneczyzewski replied:
Good call on maybe changing the name and adding comments. The trick here is it isn't only just write, it could be remove, or nothing (no-op). But I can think more on it to try and clear up the confusion. And it is complicated by the fact that what it refers to is a tricky notion (the On thing)... 🤔

shayneczyzewski replied:
Does this help improve things any?

```haskell
-- | This type tells us what we need to do with the DbSchemaChecksumOnLastDbConcurrenceFile.
data RefreshOnLastDbConcurrenceChecksumFile
  = WriteOnLastDbConcurrenceChecksumFile
  | RemoveOnLastDbConcurrenceChecksumFile
  | IgnoreOnLastDbConcurrenceChecksumFile
```

Martinsos replied:
I think it does! File makes you realize there is a noun there, I think that is enough, and comments clears it up.

shayneczyzewski replied:
Awesome, thanks!

Martinsos commented on waspc/src/Wasp/Generator/DbGenerator/Common.hs:
```diff
@@ -60,3 +66,42 @@ dbSchemaChecksumOnLastGenerateFileInDbRootDir = [relfile|schema.prisma.wasp-gene
+  | RemoveOnLastDbConcurrenceChecksum
+  | IgnoreOnLastDbConcurrenceChecksum
+
+-- When we do a create-only migration, we need to remove the DbSchemaChecksumOnLastDbConcurrenceFile so they know to migrate.
+-- When we migrate, we need to write it to indicate the local code and DB are in sync.
```
In comments you are describing imlementation. Instead, would be better to define what we expect the function to do. I would go in the direction of explaining that this function, based on the args given, decides what we should do with the checksum file. Although, this should actually be visible from the type signature. So maybe we can improve the names there and avoid the comment all together!

shayneczyzewski replied:
Good point, I can make it more about what we are doing at a higher level, and possibly remove it altogether if the sig can be cleaned up some to convey it! 👍🏻 It tends to be my crutch for, "this feels a bit complicated, just comment it to death", but that usually means the code itself can be improved to avoid it instead hehe

shayneczyzewski replied:
The names are getting longer :D but more consistent now. I did feel the comment inside does help the reader some. Do you think this is any better?

```haskell
getOnLastDbConcurrenceChecksumFileRefreshAction :: MigrateArgs -> RefreshOnLastDbConcurrenceChecksumFile
getOnLastDbConcurrenceChecksumFileRefreshAction migrateArgs =
  -- Since a create-only migration allows users to write any SQL, we remove the file to force
  -- revalidation with the DB. If it is a regular migration, we write it since they will be in sync.
  if _isCreateOnlyMigration migrateArgs
    then RemoveOnLastDbConcurrenceChecksumFile
    else WriteOnLastDbConcurrenceChecksumFile
```

Martinsos replied:
I think this looks good!

Martinsos reviewed (commented):
@shayneczyzewski I didn't have time to check it all yet, but here is first half and I will do the second one tomorrow. But this is already somewhat standalone.

shayneczyzewski commented:
> @shayneczyzewski I didn't have time to check it all yet, but here is first half and I will do the second one tomorrow. But this is already somewhat standalone.

Thanks for the first round review @Martinsos, I appreciate it! I have addressed all comments (in some form). Have a great weekend and thanks again.

Martinsos reviewed (approved):
@shayneczyzewski left some minor comments but all is looking good really, so approved! Nice job!!!

shayneczyzewski commented on waspc/test/Generator/MockWriteableMonad.hs:
```diff
@@ -77,6 +77,9 @@ instance WriteableMonad MockWriteableMonad where
   copyDirectoryRecursive srcPath dstPath = MockWriteableMonad $ do
     modifyLogs (copyDirectoryRecursive_addCall (castDir srcPath) (castDir dstPath))
 
+  removeDirectoryRecursive dir = MockWriteableMonad $ do
```
Oh yeah lol

shayneczyzewski commented on web/docs/cli.md:
```diff
@@ -16,6 +16,7 @@ COMMANDS
   GENERAL
     new <project-name>    Creates new Wasp project.
     version               Prints current version of CLI.
+    waspls                Run Wasp Language Server. Add --help to get more info.
```
Thanks! Agreed, unsure without our special doc generation app :D

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Operations.hs:
```diff
@@ -116,3 +133,19 @@ doesSchemaMatchDb genProjectRootDirAbs = do
+-- Returns Just True if all migrations are applied. Due to the fact the command
+-- returns an error on connection or unapplied migrations, Just False is never returned.
+-- It is recommended to call this after some check that confirms DB connectivity, like `doesSchemaMatchDb`.
+areAllMigrationsAppliedToDb :: Path' Abs (Dir ProjectRootDir) -> IO (Maybe Bool)
+areAllMigrationsAppliedToDb genProjectRootDirAbs = do
+  chan <- newChan
+  (_, dbExitCode) <-
+    concurrently
+      (readJobMessagesAndPrintThemPrefixed chan)
+      (DbJobs.migrateStatus genProjectRootDirAbs chan)
+  case dbExitCode of
+    ExitSuccess -> return $ Just True
+    ExitFailure _ -> return Nothing
```
Yeah this is tricky. I sorta liked it _not_ being just two states to explicitly capture there is some ambiguity in how the command returns the status. And it is similar to another command that does indicate all 3 states. My hope in the future is prisma takes my feedback and makes all their commands able to have these 3 states, so it was: a) documentation of the ambiguity, b) consistency with the one above, and c) hopefully planning ahead for the fix. For those reasons, I think it may make sense to leave as is, sound ok?

Martinsos replied:
Sounds good!

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -50,6 +50,13 @@ migrateDev projectDir maybeMigrationName = do
 
   runNodeCommandAsJob serverDir "script" scriptArgs J.Db
 
+asArgs :: MigrateArgs -> [String]
```
Will update thanks!

Martinsos commented:
LGTM!
